File: src/main/java/rr/language/psi/references/RebelFactionDescrReference.java
package rr.language.psi.references;

import com.intellij.openapi.util.TextRange;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiReference;
import com.intellij.psi.PsiReferenceBase;
import com.intellij.util.IncorrectOperationException;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import rr.language.RRUtil;
import rr.language.psi.RRRebelFactionDescrRef;

import java.util.stream.Collectors;

public class RebelFactionDescrReference extends PsiReferenceBase<PsiElement> implements PsiReference {

    public RebelFactionDescrReference(@NotNull PsiElement element, TextRange rangeInElement) {
        super(element, rangeInElement);
    }

    @Override
    public @Nullable PsiElement resolve() {
        var items = RRUtil.findAllRebelFactions(myElement.getProject()).stream()
            .map(it -> it.getRebelFactionDescrDef())
            .filter(it -> it.getText().equals(myElement.getText()))
            .collect(Collectors.toList());

        if (items.isEmpty()) {
            return null;
        }

        return items.get(0);
    }

    @Override
    public PsiElement handleElementRename(String newName) throws IncorrectOperationException {
        ((RRRebelFactionDescrRef) myElement).setName(newName);
        return myElement;
    }

    public Object @NotNull [] getVariants() {
        return new String[0];
    }
}

File: src/main/java/rr/language/psi/references/RebelFactionDescrReference.kt
package rr.language.psi.references

import com.intellij.codeInsight.lookup.LookupElementBuilder
import com.intellij.openapi.util.TextRange
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiReference
import com.intellij.psi.PsiReferenceBase
import com.intellij.util.IncorrectOperationException
import rr.language.RRUtil
import rr.language.psi.RRRebelFactionDescrRef

class RebelFactionDescrReference(element: PsiElement, rangeInElement: TextRange?) :
    PsiReferenceBase<PsiElement?>(element, rangeInElement), PsiReference {
    override fun resolve(): PsiElement? {
        val items = RRUtil.findAllRebelFactions(myElement!!.project)
            .map { it.rebelFactionDescrDef }
            .filter { it.text == myElement!!.text }

        return items.firstOrNull()
    }

    @Throws(IncorrectOperationException::class)
    override fun handleElementRename(newName: String): PsiElement {
        (myElement as RRRebelFactionDescrRef).setName(newName)
        return myElement as RRRebelFactionDescrRef
    }

    override fun getVariants(): Array<Any> {
        return RRUtil.findAllRebelFactionsAsStrings(myElement!!.project)
            .map { LookupElementBuilder.create(it) }
            .toTypedArray()

    }
}
